### PR TITLE
Improve TimeVal

### PR DESCRIPTION
**Pull request by elmarco**
Make it a new type and implement a few useful traits to pass it as
argument to glib functions.

Signed-off-by: Marc-André Lureau <marcandre.lureau@redhat.com>

### CODE REVIEW

File: src/time_val.rs [hunk 1]
```diff
@@ -4,13 +4,63 @@
...
+        let ret = unsafe { mem::zeroed() };
+        TimeVal(ret)
+    }
+}
+
+impl fmt::Debug for TimeVal {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("TimeVal")
+            .field("tv_sec", &self.0.tv_sec)
+            .field("tv_nsec", &self.0.tv_usec)
```
Comment (sdroege) [inline src/time_val.rs]: You should add setters/getters for these fields also, or otherwise it would be a bit useless

**Comment by GuillaumeGomez** [comment]
👍

**Comment by EPashkin** [comment]
No changes to Gir.toml, so CI will fail on checking generated code in "src/auto/date_time.rs"

File: src/time_val.rs (hunk 1 above)
Comment (elmarco) [reply src/time_val.rs]: ok, after a second thought, I drop the newtype for now.

**Comment by elmarco** [comment]
@EPashkin what should I change?

**Comment by EPashkin** [comment]
@elmarco Add "GLib.TimeVal" to "manual" array, then run regenerate code with "make gir"
```diff
+++ b/Gir.toml
@@ -32,6 +32,7 @@ ignore = [
 manual = [
     "GLib.Bytes",
     "GLib.Error",
+    "GLib.TimeVal",
     "GLib.Variant",
```

File: src/auto/date_time.rs [hunk 2]
```diff
@@ -265,9 +270,11 @@ impl DateTime {
         }
     }
 
-    //pub fn to_timeval(&self, tv: /*Ignored*/&mut TimeVal) -> bool {
-    //    unsafe { TODO: call glib_sys:g_date_time_to_timeval() }
-    //}
+    pub fn to_timeval(&self, tv: &mut TimeVal) -> bool {
```
Comment (sdroege) [inline src/auto/date_time.rs]: This should be `pub fn to_timeval(&self) -> Option<TimeVal>`

**Comment by sdroege** [comment]
Looks good to me otherwise, thanks

File: src/auto/date_time.rs (hunk 2 above)
Comment (elmarco) [reply src/auto/date_time.rs]: as manual binding?

Comment (sdroege) [reply src/auto/date_time.rs]: I'm not sure why gir doesn't autogenerate it as such. @EPashkin ?

But yes, manual binding until gir is fixed

Comment (GuillaumeGomez) [reply src/auto/date_time.rs]: A new bug has been found! Yeay!

Comment (GuillaumeGomez) [reply src/auto/date_time.rs]: Release just got delayed by another bug. :3

**Comment by sdroege** [comment]
@elmarco What's the status of this?

**Comment by elmarco** [comment]
I haven't looked at the to_timeval() generation issue

**Comment by sdroege** [comment]
> I haven't looked at the to_timeval() generation issue

It's ok to implement it manually for now and simply create an issue against gir.

**Comment by sdroege** [comment]
Please rebase and create a new PR for https://github.com/gtk-rs/gtk-rs . Thanks!

**Pull request closed without merging**